services/leader/client.go

```diff
@@ -0,0 +1,229 @@
+	CampaignError
+
+	// CampaignClosed indicates the campaign has been closed.
+	CampaignClosed
+)
+
+// appended to elections with an empty string for electionID to make it easier
+// for user to debug etcd keys
+const defaultElectionSuffix = "default"
+
+var (
+	// ErrSessionExpired is returned when a client's session (etcd lease) is no
+	// longer being refreshed for any reason (due to expiration, error state,
+	// etc.).
+	ErrSessionExpired = errors.New("election client session (lease) expired")
+)
+
+// NB(mschalle): when an etcd leader failover occurs, all current leases have
+// their TTLs refreshed: https://github.com/coreos/etcd/issues/2660
+
+type client struct {
+	sync.Mutex
+
+	ctxCancel   context.CancelFunc
+	election    *concurrency.Election
+	session     *concurrency.Session
+	val         string
+	closed      uint32
+	campaigning uint32
+	wb          xwatch.Watchable
+}
+
+// newClient returns an instance of an client client bound to a single election.
+//
+// TODO(mschalle): condense all the places options can be passed + allow caller
+// to set TTL opts
+func newClient(cli *clientv3.Client, opts Options, electionID string) (*client, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	var sessionOpts []concurrency.SessionOption
+	if ttl := opts.TTL(); ttl != 0 {
+		sessionOpts = append(sessionOpts, concurrency.WithTTL(ttl))
+	}
+
+	session, err := concurrency.NewSession(cli, sessionOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	election := concurrency.NewElection(session, electionPrefix(opts.ServiceID(), electionID))
+
+	wb := xwatch.NewWatchable()
+	wb.Update(CampaignUnstarted)
+
+	return &client{
+		election: election,
+		session:  session,
+		val:      opts.OverrideValue(),
+		wb:       wb,
+	}, nil
+}
+
+func (c *client) Campaign() (xwatch.Watch, error) {
+	if c.isClosed() {
+		return nil, ErrClientClosed
+	}
+
+	if !atomic.CompareAndSwapUint32(&c.campaigning, 0, 1) {
+		return nil, ErrCampaignInProgress
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	c.Lock()
+	c.ctxCancel = cancel
+	c.Unlock()
+
+	go func() {
+		c.wb.Update(CampaignFollower)
+		// blocks until elected or error
+		err := c.election.Campaign(ctx, c.val)
+		if err != nil {
+			c.wb.Update(err)
+		}
+
+		select {
+		case <-c.session.Done():
+			c.wb.Update(ErrSessionExpired)
+		default:
+			c.wb.Update(CampaignLeader)
+		}
+
+		cancel()
+		c.Lock()
+		c.ctxCancel = nil
+		c.Unlock()
+	}()
+
+	_, w, err := c.wb.Watch()
+	if err != nil {
+		return nil, err
+	}
+
+	return w, nil
+}
+
+func (c *client) Resign() error {
+	if c.isClosed() {
+		return ErrClientClosed
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	// if we're not the leader but still campaigning, cancelling the context
+	// will stop the campaign
+	c.cancelWithLock()
+
+	// TODO(mschalle): resign context timeout?
+	if err := c.election.Resign(context.Background()); err != nil {
+		c.wb.Update(err)
+		return err
+	}
+
+	c.wb.Update(CampaignFollower)
```

Comment:
I would argue so. If you're not the leader but you haven't closed your client then follower sounds appropriate. I'll change the default state from `Unstarted` to `Follower` to reflect this.